processor/src/main/java/org/mapstruct/ap/internal/processor/Jsr330ComponentProcessor.java

```diff
@@ -42,7 +42,7 @@ protected String getComponentModelIdentifier() {
     @Override
     protected List<Annotation> getTypeAnnotations(Mapper mapper) {
         if ( mapper.getDecorator() == null ) {
-            return Collections.singletonList( named() );
+            return Arrays.asList( singleton(), named() );
         }
         else {
             return Collections.singletonList( namedDelegate( mapper ) );
```

Comment:
You need to add the annotation here as well, as otherwise the delegate (i.e. the actual mapper implementation) for a decorator does not get the `@Singleton` annotation.